### PR TITLE
Add Battery Stats

### PR DESCRIPTION
**Pull request by klahnakoski**

### CODE REVIEW

**Comment by klahnakoski** [comment]
@gmierz please review https://deploy-preview-431--firefox-health-dashboard.netlify.com/power?platform=p2-aarch64&suite=speedometer

**Comment by klahnakoski** [comment]
@gmierz  

* What is `scn-power-idle`?  is that a test, or is it an idle machine?  What does the "scn" stand for?
* Are there other test suites planned?

**Comment by gmierz** [comment]
> 
> 
> @gmierz
> 
>     * What is `scn-power-idle`?  is that a test, or is it an idle machine?  What does the "scn" stand for?

This is an idle test with the browser running on the about:blank page. "scn" stands for scenario which is the test type.

> 
>     * If it is idle, then why is the power consumption higher?

This is because a plain white background (the about:blank page) consumes the most energy on the P2 devices, so it uses more than the speedometer test.

> 
>     * Are there other test suites planned?
>

Yes, we will be adding more tests such as `scn-power-idle-bkg`.

And I think the interface looks good, nice work! I have the following comments:
1) It is unclear what test suite a certain color is a part of (in the data points).
2) The solid dots don't follow each other properly when browsing the data points. When multiple suites are overlayed, the solid dots seems to start at the beginning (time-wise) of each of the time-series.
3) It could be useful to have the ability to overlay test suites across platforms (not necessary).

**Comment by klahnakoski** [comment]
>     1. It is unclear what test suite a certain color is a part of (in the data points).

Fixed.  I will show only one suite at a time now.

>     2. The solid dots don't follow each other properly when browsing the data points. When multiple suites are overlayed, the solid dots seems to start at the beginning (time-wise) of each of the time-series.

Yes, this is a known problem:  https://github.com/mozilla-frontend-infra/firefox-health-dashboard/issues/386

>     3. It could be useful to have the ability to overlay test suites across platforms (not necessary).

@gmierz 

I agree.  It would be nice if the user can choose how to show the four dimensions (`Platform` x `Browser` x `Suite` x `Test`) on the 2-dimentional page (`series-in-chart`, `chart-on-page`).  Right now the `series-in-chart` is showing `Browser`, which is uninteresting.  Maybe it should show `Platform`, and leave the user to select `Browser` at the top?

**Comment by klahnakoski** [comment]
@gmierz I pushed a new version that shows all platforms.  Which do you like?

 https://deploy-preview-431--firefox-health-dashboard.netlify.com/power?browser=geckoview&suite=speedometer

File: src/components/DashboardPage/index.jsx [hunk 1]
```diff
@@ -30,15 +38,25 @@ const DashboardPage = ({ classes, children, title, subtitle }) => (
...
+        <Link to="/quantum/64" title="Quantum 64 Performance">
+          <DesktopIcon />
+        </Link>
+        <Link to="/power" title="Power Usage">
+          <BatteryIcon />
+        </Link>
+        <a
+          href="https://github.com/mozilla-frontend-infra/firefox-health-dashboard/"
+          title="Source Code">
+          <CodeIcon />
```
Comment (klahnakoski) [inline src/components/DashboardPage/index.jsx]: above code is showing icons in top left of screen:

image

File: src/components/Section/index.jsx [hunk 2]
```diff
@@ -29,14 +30,15 @@ const Section = ({ classes, children, subtitle, title, more }) => (
...
+          {subtitle && <span className={classes.subtitle}>{subtitle}</span>}
           {more && (
             <span>
-              {' ('}
-              <a href={more}>more</a>
-              {')'}
+              {' '}
+              <a href={more} title="show details">
+                <DetailsIcon />
+              </a>
```
Comment (klahnakoski) [inline src/components/Section/index.jsx]: Use icon instead of word

File: src/components/TP6mAggregate.jsx [hunk 3]
```diff
@@ -287,12 +287,11 @@ class TP6mAggregate_ extends Component {
                                 platform,
                               },
                             })}
+                            title="show details"
                             target="_blank"
-                            rel="noopener noreferrer"
-                            style={{ fontSize: '0.9em' }}>
-                            {'details'}
+                            rel="noopener noreferrer">
+                            <DetailsIcon />
```
Comment (klahnakoski) [inline src/components/TP6mAggregate.jsx]: Use icon instead of word

File: src/containers/PerfherderGraphContainer/index.jsx [hunk 4]
```diff
@@ -20,15 +20,15 @@ const ALLOWED_TREEHERDER_TIMERANGES = [1, 2, 7, 14, 30, 60, 90].map(
...
 
   return {
-    reverse: !lowerIsBetter,
     tooltips: {
       enabled: false,
     },
     // Start using: chartSchema.md
-    'axis.y.label': lowerIsBetter ? 'Duration' : 'Score',
+    'axis.y.label': unit,
+    'axis.y.reverse': !lowerIsBetter,
```
Comment (klahnakoski) [inline src/containers/PerfherderGraphContainer/index.jsx]: fixing the perherder data to emit `unit`, and we can use it here

File: src/containers/PerfherderGraphContainer/index.jsx [hunk 5]
```diff
@@ -132,10 +132,6 @@ const styles = () => ({
     padding: '.2rem .3rem .3rem .3rem',
     margin: '0 1rem 0 0',
   },
-  linkIcon: {
-    marginLeft: '0.2rem',
-    marginBottom: -5,
-  },
```
Comment (klahnakoski) [inline src/containers/PerfherderGraphContainer/index.jsx]: remove CSS where it is obvious we do not need it

File: src/containers/RedashContainer/index.jsx [hunk 6]
```diff
@@ -81,32 +82,29 @@ class RedashContainer extends Component {
...
+              {title}{' '}
+              <a
+                href={redashQueryUrl}
+                target="_blank"
+                rel="noopener noreferrer"
+                title="go to Redash query">
+                <LinkIcon />
+              </a>
+            </div>
+          }
```
Comment (klahnakoski) [inline src/containers/RedashContainer/index.jsx]: Use icon instead of words

File: src/index.css [hunk 7]
```diff
@@ -12,19 +12,18 @@
...
-  height:100%
+ html, body {
+  overflow-x: hidden;
+  position: absolute;
+  top: 0;
+  bottom: 0;
+  left: 0;
+  right: 0;
+  margin: 0;
+  padding: 0;
```
Comment (klahnakoski) [inline src/index.css]: There is zero-pixel overflow, which triggered the horizontal scrollbar, which is now gone.

File: src/index.css [hunk 8]
```diff
@@ -46,49 +45,10 @@
...
-  text-align:right
- }
- .dashboard-title .dashboard-source a:last-child {
-  display:block
- }
- .dashboard-title .dashboard-link {
-  padding:.53572rem .375rem;
-  color:#00a7e0;
-  display:none
- }
```
Comment (klahnakoski) [inline src/index.css]: more CSS removal

File: src/power/config.js [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+    ['geckoview', 'Geckoview', 'speedometer', 'Speedometer', {eq: {framework: 10, repo: 'mozilla-central', suite: ['raptor-speedometer-geckoview-power']}}],
+    ['refbrow', 'Reference Browser', 'speedometer', 'Speedometer', {eq: {framework: 10, repo: 'mozilla-central', suite: ['raptor-speedometer-refbrow-power']}}],
+    ['fenix', 'Firefox Preview', 'speedometer', 'Speedometer', {eq: {framework: 10, repo: 'mozilla-central', suite: ['raptor-speedometer-fenix-power']}}],
+    ['fennec', 'Fennec', 'speedometer', 'Speedometer', {eq: {framework: 10, repo: 'mozilla-central', suite: ['raptor-speedometer-fennec-power']}}],
+
+  ],
+};
+const COMBOS = COMBO_TABLE.data.map(row => Data.zip(COMBO_TABLE.header, row));
+
+export { TESTS, COMBOS, PLATFORMS };
```
Comment (klahnakoski) [inline src/power/config.js]: This is the config file to drive the `/power` page

File: src/power/index.jsx [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+    id: 'suite',
+    label: 'Suite',
+    defaultValue: 'speedometer',
+    options: selectFrom(COMBOS)
+      .groupBy('suiteLabel')
+      .map(([v]) => ({ id: v.suite, label: v.suiteLabel })),
+  },
+];
+
+export default withNavigation(nav)(withStyles(styles)(Power));
```
Comment (klahnakoski) [inline src/power/index.jsx]: The actual `/power` page

File: src/quantum/Quantum.jsx [hunk 11]
```diff
@@ -581,11 +582,10 @@ export default class QuantumIndex extends React.Component {
...
                 {more && (
                   <span>
-                    {' ('}
-                    <a href={more} style={{ fontSize: '0.8em' }}>
-                      more
+                    {' '}
+                    <a href={more} title="show details">
+                      <DetailsIcon />
                     </a>
-                    {')'}
```
Comment (klahnakoski) [inline src/quantum/Quantum.jsx]: Use icon instead of word

File: src/quantum/config.js [hunk 12]
```diff
@@ -47,7 +44,6 @@ const CONFIG = {
       title: 'MotionMark Animometer',
       secondLink:
         '/quantum/10/windows7-32/raptor-motionmark_animometer-firefox/pgo',
-      secondTitle: 'Breakdown',
```
Comment (klahnakoski) [inline src/quantum/config.js]: we no longer need the word because we use icon

File: src/quantum/widget.jsx [hunk 13]
```diff
@@ -35,14 +36,14 @@ export default class Widget extends React.Component {
...
-        $secondTitle = this.props.secondTitle;
-      }
+    if (this.props.secondLink) {
+      $secondTitle = (
+        <Link to={this.props.secondLink} title="show details">
+          <DetailsIcon />
+        </Link>
+      );
+    } else {
+      $secondTitle = '';
```
Comment (klahnakoski) [inline src/quantum/widget.jsx]: Use icon instead of word

File: src/utils/icons.jsx [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+  TimerIcon,
+  BatteryIcon,
+  VideoIcon,
+  DesktopIcon,
+  AndroidIcon,
+  DetailsIcon,
+  GithubIcon,
+  CodeIcon,
+  HomeIcon,
+};
```
Comment (klahnakoski) [inline src/utils/icons.jsx]: All icons are here, plus some styling because they are badly behaved

File: src/vendor/chartJs/utils.js [hunk 15]
```diff
@@ -74,7 +75,7 @@ const generateOptions = (rawOptions = {}, data) => {
     {
       ticks: {
         beginAtZero: true,
-        reverse,
+        reverse: Data.get(options, 'axis.y.reverse'),
```
Comment (klahnakoski) [inline src/vendor/chartJs/utils.js]: FIX: be sure to pass in the `reverse`

File: src/vendor/logs.js [hunk 16]
```diff
@@ -25,25 +25,28 @@ function parseStack(stackString) {
...
+            fileName: parts[2],
+            line: parts[3],
+            column: parts[4],
+          };
+        })
+        .exists()
+        .first()
+    )
+    .exists()
+    .toArray();
```
Comment (klahnakoski) [inline src/vendor/logs.js]: A little more scrubbing of the stack traces

File: src/vendor/perfherder.js [hunk 17]
```diff
@@ -40,10 +40,23 @@ const getFramework = async combo => {
...
-          let lowerIsBetter = true;
+          const cleanTest = (() => {
+            if (missing(test)) return null;
+
+            if (test === suite) return null;
+
+            if (test.startsWith(suite)) return test.slice(suite.length + 1);
+
+            return test;
+          })();
```
Comment (klahnakoski) [inline src/vendor/perfherder.js]: Test names are can come in many different flavors: We scrub them.

File: src/vendor/perfherder.js [hunk 18]
```diff
@@ -40,10 +40,23 @@ const getFramework = async combo => {
...
+          let lowerIsBetter = lower_is_better;
+          let unit = 'Score';
 
           if (lower_is_better === undefined) {
-            if (
+            if (suite.endsWith('-power')) {
+              lowerIsBetter = true;
+              unit = 'mAh';
+            } else if (
```
Comment (klahnakoski) [inline src/vendor/perfherder.js]: Markup `power` tests with correct unit

File: src/vendor/utils/links.jsx [hunk 19]
```diff
@@ -0,0 +1,21 @@
...
+    return (
+      <a href={to} title={title}>
+        {children}
+      </a>
+    );
+  }
+}
+
+/* eslint-disable-next-line import/prefer-default-export */
+export { Link };
```
Comment (klahnakoski) [inline src/vendor/utils/links.jsx]: The builtin `Link` can not be used outside a router, which then breaks the tests.  This `Link` does not have that problem.

File: src/views/Android/TP6m.jsx [hunk 20]
```diff
@@ -130,69 +127,65 @@ class TP6M extends React.Component {
       .first().label;
 
     return (
-      <div className={classes.body}>
```
Comment (klahnakoski) [inline src/views/Android/TP6m.jsx]: remove more styling

File: src/views/Home/index.jsx [hunk 21]
```diff
@@ -35,10 +36,18 @@ const Home = ({ classes }) => (
...
+      </Link>
+      <Link to="/quantum/32">
+        <DesktopIcon /> Quantum 32bit
+      </Link>
+      <Link to="/quantum/64">
+        <DesktopIcon /> Quantum 64bit
+      </Link>
+      <Link to="/battery">
+        <BatteryIcon /> Power Usage
+      </Link>
```
Comment (klahnakoski) [inline src/views/Home/index.jsx]: Add link to `/battery` and add icons

File: src/routes.jsx [hunk 22]
```diff
@@ -18,13 +18,14 @@ const Routes = () => (
...
       <Route path="/android/graph" component={NimbledroidGraphPage} />
+
       <Route path="/android" component={AndroidPage} />
+      <Route path="/power" component={Power} />
       <Route
         path="/quantum/:framework/:platform/:suite/:option"
         component={Subbenchmark}
       />
       <Route path="/quantum/:bits" component={Quantum} />
-      <Route path="/js-team" component={JsTeam} />
```
Comment (klahnakoski) [inline src/routes.jsx]: removed `js-team` page because it was never done

**Comment by klahnakoski** [comment]
@sarah-clements   Please review this PR. 

There important part is the addition of the `power` detail page:

* [power/config.js](https://github.com/mozilla-frontend-infra/firefox-health-dashboard/pull/431/files#diff-a4b283cd94bdf6fe52515d6d43a065b4)
* [power/index.jsx](https://github.com/mozilla-frontend-infra/firefox-health-dashboard/pull/431/files#diff-9259b06913a0987519c654f50685f409)

Some other notes:

1. The rest is CSS cleanup, and adding clickable icons for menus, tabs, and drill-down.  I added comments to explain the various changes.  
2. I must still discuss with management about what they want to see on the main dashboard: That will be done in another PR.  

Thank you

**Comment by klahnakoski** [comment]
@karlht  Your review is also appreciated

**Comment by gmierz** [comment]
@klahnakoski I prefer the second one that let's us see all the platforms for a given test suite. But perf-sheriffs might prefer the original interface so it might be a good idea to ask igoldan about it as well. I think this depends on how we will look at the data to locate regressions, and this will constantly change from one issue to another.

**Comment by klahnakoski** [comment]
> @klahnakoski I prefer the second one that let's us see all the platforms for a given test suite. But perf-sheriffs might prefer the original interface so it might be a good idea to ask igoldan about it as well. I think this depends on how we will look at the data to locate regressions, and this will constantly change from one issue to another.

@gmierz   This dashboard is not meant for performance sheriffing, nor for developers; The audience is managers: They will be providing their perspective this week, so it may change again.

File: src/components/DashboardPage/index.jsx (hunk 1 above)
Comment (sarah-clements) [reply src/components/DashboardPage/index.jsx]: Icons look nice but what does the <> mean? Perhaps you could find a clearer icon?

File: src/power/config.js [hunk 23]
```diff
@@ -0,0 +1,74 @@
+/* eslint-disable */
+
+import { Data } from '../vendor/Data';
+
+const TESTS = [
+  {
+    id: 'cpu',
+    label: 'CPU',
+    filter: { eq: { test: 'cpu' } },
```
Comment (sarah-clements) [inline src/power/config.js]: It seems like for something like an id (presumably used for internal purposes) you can just use the label?

File: src/quantum/widget.jsx (hunk 13 above)
Comment (sarah-clements) [reply src/quantum/widget.jsx]: I'm curious why you're using a $ in front of secondTitle.

File: src/utils/icons.jsx [hunk 24]
```diff
@@ -0,0 +1,60 @@
+import React from 'react';
+import LinkIco from '@material-ui/icons/Launch';
+import HomeIco from '@material-ui/icons/Home';
+import TimerIco from '@material-ui/icons/Timer';
+import BatteryIco from '@material-ui/icons/BatteryChargingFull';
+import VideoIco from '@material-ui/icons/Videocam';
+import DesktopIco from '@material-ui/icons/DesktopWindows';
+import AndroidIco from '@material-ui/icons/PhoneAndroid';
+import DetailsIco from '@material-ui/icons/MoreHoriz';
+import CodeIco from '@material-ui/icons/Code';
```
Comment (sarah-clements) [inline src/utils/icons.jsx]: It's odd to leave the 'n' off of Icon here. I think it's better to have a longer name that's clearer, rather than a short vague one :) So `materialLinkIcon` or or rename your icon components to something like `LinkIconStyle`.

However, since some of the styling is the same, couldn't you create a function that returns the div with the style and pass in the Icon element or prop (for `LinkIcon` and `DetailsIcon`)?

File: src/quantum/widget.jsx (hunk 13 above)
Comment (klahnakoski) [reply src/quantum/widget.jsx]: That is curious; I did not even notice!  I was simply using the variable that was used before.  I will fix this.

File: src/power/config.js (hunk 23 above)
Comment (klahnakoski) [reply src/power/config.js]: In general, I separate the machine-friendly `id` from the humane `label` because they too-often diverge. 

Humans like spaces and capitalization, and sometimes like names to change, while machines like consistency (eg all lowercase, or not changing) and standardization (eg use third party names, like from perfherder).  For example; I have no idea what `proportional` means: I will learn shortly, and it will be easy to change.

File: src/utils/icons.jsx (hunk 24 above)
Comment (klahnakoski) [reply src/utils/icons.jsx]: Yes, since these names do not escape this module, I did not care to invent a better name.  The linter gets angry with any names with prefix or suffix of underscore (`_`), so instead of adding a character, I removed one. I will add `Internal` as a suffix instead.

> However, since some of the styling is the same, couldn't you create a function that returns the div with the style and pass in the Icon element or prop (for `LinkIcon` and `DetailsIcon`)?

I think what you suggest will be just as much code as what I have done.

Comment (sarah-clements) [reply src/utils/icons.jsx]: I was thinking something like this :) But maybe not worth for two icons.
```
const IconWithStyle = (icon) => (
  <div
    style={{
      display: 'inline-block',
      margin: '-1rem 0',
      verticalAlign: 'middle',
    }}>
    {icon}
  </div>
);
```

Comment (klahnakoski) [reply src/utils/icons.jsx]: @sarah-clements 

I have a question about your suggestion:  Is styling an outer `div` that contains an inner component the same as styling the inner component? Since CSS is capable of insane inter dependencies; I am worried that more html tags, means more CSS styling opportunities, which results in more CSS rat-nest logic. This assumes I can even get the inner component to behave by styling the outer `div`.

On the other hand I have not seen this nightmare scenario in real life, so I wonder if there is some React magic happening here.

Comment (sarah-clements) [reply src/utils/icons.jsx]: I think this suggestion would be the same as embedding the component - you're styling the outer container and then any props passed directly to the component before its passed in would apply to that component's container (or whatever its used for).
